src/tactic/move_add.lean

```diff
@@ -0,0 +1,378 @@
+/-
+Copyright (c) 2022 Arthur Paulino, Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Arthur Paulino, Damiano Testa
+-/
+import tactic.core
+import algebra.group.basic
+
+/-!
+# `move_add`: a tactic for moving summands
+
+Calling `move_add [a, ← b, c]`, recursively looks inside the goal for expressions involving a sum.
+Whenever it finds one, it moves the summands that unify to `a, b, c`, removing all parentheses.
+
+See the doc-string for `tactic.interactive.move_add` for more information.
+
+##  Implementation notes
+
+This file defines a general `move_op` tactic, intended for reordering terms in an expression
+obtained by repeated applications of a given associative, commutative binary operation.  The
+user decides the final reordering.  Applying `move_op` without specifying the order will simply
+remove all parentheses from the expression.
+The main user-facing tactics are `move_add` and `move_mul`, dealing with addition and
+multiplication, respectively.
+
+In what is below, we talk about `move_add` for definiteness, but everything applies
+to `move_mul` and to the more general `move_op`.
+
+The implementation of `move_add` only moves the terms specified by the user (and rearranges
+parentheses).
+
+Note that the tactic `abel` already implements a very solid heuristic for normalizing terms in an
+additive commutative semigroup and produces expressions in more or less standard form.
+The scope of `move_add` is different: it is designed to make it easy to move individual terms
+around a sum.
+
+##  Future work
+
+* Add support for `neg/div/inv` in additive/multiplicative groups?
+* Add operations other than `+` and `*`?  E.g. `∪, ∩, ⊓, ⊔, ...`?
+  Should there be the desire for supporting more operations, it might make sense to extract
+  the `simp [add] <|> simp [mul]` block in `reorder_hyp` to a separate tactic,
+  including all the lemmas used for the rearrangement to work.
```

Comment:
Can we just use `ac_refl` for this? I understand it's much slower than using the `simp`s, but we can keep those as optimizations for the `+` and `*` cases.

Reply:
Is the suggestion to match on the operation, using `simp [whatever]` for add and mul and `ac_refl` otherwise?  I like this idea!  We could also directly support intersections and unions, as it is hardly any more work...

Reply:
I was suggesting the much worse approach of `simp only [add_things] <|> simp only [mul_things] <|> ac_refl`, since I think the latter knows how to do pretty much everything else. The other approach would be to lookup instances of `` `(is_commutative _ %%op)`` and `` `(is_associative _ %%op)`` and then invoke `simp` with lemmas stated in  terms of those typeclasses, which I asume exist somewhere.

Reply:
I had also thought that we could create a custom typeclass whose fields are op_assoc and op_comm, prove op_left_comm and then simply add instances for all operations that we want to support.

Reply:
I just pushed a version that uses `simp [add-lemmas]` for addition, `simp [mul-lemmas]` for multiplication and `ac_refl` otherwise.  I also tried having a general tactic taking the operation as input.

I see two main issues:
1. the operation must be given inside square brackets, since I do not know how to make the parser stop parsing;
2. the `ac_refl` branch only actually works on an equality and does not recurse inside the expression,